### ISSUE TITLE
UI recording and UI bytecode interpreter

### ISSUE THREAD

**Issue by vurtun** [vurtun/zahnrad#59]
As some may know I added the functionality to `record` UI and `replay` it at a later point of time some time ago. The main premise is that since the recorded UI is fixed allows for a lot more control like the ability to check if the UI needs to redrawn or even a lot more widget depended information. Internally this is achieved by serializing function call and parameter into a buffer that later can be executed again. State changes are signaled in contrast to the typical immediate mode API by using events while also providing an API to change the serialized widget state. So in the end not much is lost while a lot is achieved.

So far this would already allow things like an editor comparable to Qt but more powerful since dynamic widgets like combo boxes, charts or any highly dynamic UI code which are tedious to work with in classic retain mode APIs could be generated by recording or using the immediate mode API. So in theory this alone could be a huge win.

I thought about it a little more and I noticed that the step between just serializing widgets and a custom script language is not far fetched. I am not talking about bindings to other languages which can already be done, but something more aligned to shader languages in graphics APIs. The basic idea would be that the language (If it even could called that since it probably does not have to be turing complete) would call different widget just like you would with the default immediate mode API   (which will still be usable like before) but would generate custom events as output (since you cannot directly access your program). So far example this script:

Script (pseudo code):
```c
event changed_mouse_sensitivity {
    float value;
}
event mouse_sensitivity_greater_two {
    float value;
}

void main()
{
    s = slider("mouse_sensitivity", 0, 5, 10);
    if (s.isModified()) {
        event changed_mouse_sensitivity evt;
        evt.value = s.value;
        generateEvent(evt);
    }
    if (s.value > 2) {
        event mouse_sensitivity_greater_two evt;
        evt.value = s.value;
        generateEvent(evt);
    }
}
```

Native code:
```c
/* register custom event types here */
zr_exec(...); /* run 'script' */

/* first way to check for changes */
zr_foreach_event(....) {
    if (evt.type == CUSTOM_EVENT_CHANGE_MOUSE_SENSITIVITY) {
        /* handle mouse sensitivity change */
    }      
}

/* second way to access and check for UI changes directly */
zr_element mouse_sensitivity = zr_lookup_string(program, "mouse_sensitivity");
if (zr_element_is_modified(mouse_sensitivity)) {
    /* handle mouse sensitivity change */
    zr_element_clear_modified(mouse_sensitivity);
}
```

So in script you would create a slider and check if the current slider value is greater than two and would generate a custom event if so. So in your native code you wouldn't have to check for each slider if it has been modified but instead you would only listen to custom events. It would be especially helpful for button events. In addition you could also access the widget by handle and check for changes directly as well. While this script could have logic like I described it could also just define each widget as data (used by editors) like so (still pseudo code):

```c 
begin("Demo")
layout_row_dynamic(30, 2);
slider("mouse_sensitivity", 0, 0.5, 1.0);
button("ok_button", "ok");
end()
```
So this would look more like the bytecode generated currently by `recording` and would just use default events for clicking, hovering, etc but could be extended by modifying the generated editor textual output to have custom events.

What I described up to this point would stand as three different APIs that would support each other. First the default immediate mode API which is the base for and runs everything. Then above that would be UI recording which allows to use the immediate mode API to turn dynamic frame based UI to a fixed bytecode UI script and with that more control and information and finally the bytecode script layer. 

These three could be used interchangeably.
So for example you have an UI which means in most cases that most of its layout is quite fixed and is not required to change a lot but you still update and redraw it every frame. Instead you could turn that into script and run that from data. 
Lets extend the use case and say your UI has a combo box which needs to be filled at the beginning of the program. In classical retain mode APIs you would have to fill a data structure and pass that to the combo box which is quite tedious for more advanced changes like removing and in using the immediate mode API you would have to call and redraw the combo box every frame. 
Now in this case the recording API would become handy since it would generate the combo box for you at the beginning of the program and could then be combined with the fixed UI script.
Finally lets extend it a third time and say we have some highly dynamic data used for charts. In this case we would use the immediate mode API.

This is basically the high level idea I had. There are of course a lot of problem related to this like how to extend the script language with custom widgets, the best way to combine recorded and normal script, what script language with what syntax, what control statements should the language support, etc. 

So finally I want to head to my question. I personally don't have much experience with languages I wrote an emulators once so I know a little bit how to write bytecode but I have no experience about if what I described could be or even should be done by using a script language. I suspect no since what the language would actually do is very limited it is barley more than just data. For example it wouldn't need functions or probably even variables and is therefore quite simple. I know for example that bitcoin uses a very basic interpreter for something related and I will take another look into that.

So for my question I would ask if what I am thinking about even made sense or is just over engineered bullshit that could be done by using a script language, should it even be done? or should the library stay simple and if it is worth it then if it should be done by only using bytecode inside the library and provide a way to generate the bytecode outside the library or since the language is so simple would even make sense to actually implement everything inside the library?
So in general feedback would be appreciated.

**Comment by fr500** [comment]
I think this is seriously nice but I'm not sure what are the use cases, I'm not even certain of what the recording feature does.
The immediate mode and simplicity is what actually got my attention and it works really well with my project.

**Comment by vurtun** [comment]
> I think this is seriously nice but I'm not sure what are the use cases, I'm not even certain of what the recording feature does.
The immediate mode and simplicity is what actually got my attention and it works really well with my project.

The immediate mode API would work like before and it would just be extended with additional functionality for other use cases. 

But in the end you are right and I spend a lot of time thinking about this and finally decided to remove recording and everything about this idea. While what I described here is probably interesting I just don't have the energy. I spend over a year on this project and I just can't get myself to commit to these big changes anymore.
I will probably also not do other bigger features I was thinking about like docking or skinning. I definitely will further fix bugs or essential must need features or utility functions if required and if I have the motivation for it will extend the internal command drawing API with a path drawing API (same as used in the canvas) to allow for more interesting custom widget drawing. 

In the end I hope this library is helpful for some and not a total waste of time for others and I will probably take a break either after extending the command drawing API or from now on and will only fix things and of course answer issues.

**Comment by dumblob** [comment]
In my humble opinion, recording is the great idea which makes this library far ahead of any other immediate mode GUI. I'm convinced, that this functionality should be provided even by a lean immediate mode library like this one earlier or later.

I can totally understand that making of this library is very demanding and time consuming. I think there is no need to hurry that much (even though it might seem opposite from the reactions of users ;) ). Take a deep look at the rapidly growing number of stars you're getting on GitHub. Consider also, that despite this high number, there are nearly no GitHub issues reported - i.e. the library has really high quality.

Last but not least, there is a good chance, that after contacting few game studios, you could get hired part time to work (remotely) on this library (I know in person few guys who did something similar) - there is only one thing which needs to be done first, namely let the studios know about this awesome library (I'll start - next week I'm meeting one good game developer, so I'll push him to spread a word about this lib among the whole company he's working at - it's a moderately known company among RPG-makers).

And by the way, it's normal, that open-source projects are slower in their development than usual projects - so one year waiting for a new mid-sized feature is a normal and absolutely legit "delay".

So, feel free to take a break and we'll hope you'll resurrect development (as opposed to only bug fixing) at some point.

**Comment by dumblob** [comment]
I forgot to add, that I was thinking about providing the recording feature as another separate "higher-level" header file instead of being part of Zahnrad as it's a little bit more advanced and possibly not needed for quick&dirty simple UI creation.

I can imagine that this separate "Zahnrad recording library" would also contain a growing set of slightly advanced widgets (a good example, as already mentioned above, would be a basic text editing widget supporting different sizes, colors, font effects, multiple selection, rotation, configurable keyboard & mouse bindings, etc.).

This way, we could also encourage those who want advanced functionality to provide their "UI shaders describing special advanced widgets" independently of the main Zahnrad library.

**Comment by fr500** [comment]
@vurtun I tried lots of toolkits and this is by far the easiest one to integrate.
I understand doing a big project like this by yourself can be a daunting task and if you feel tired you should of course take a break.

After all, this is an open-source project and we can all contribute with our changes if there are any at some point. Thanks for all your work so far.
